cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/ElasticsearchSource.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.hadoop.io.DoubleWritable;
+import org.apache.hadoop.io.FloatWritable;
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.MapWritable;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.Job;
+import org.elasticsearch.hadoop.mr.EsInputFormat;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link BatchSource} that writes data to Elasticsearch.
+ * <p>
+ * This {@link ElasticsearchSource} reads from an Elasticsearch index and type and converts the MapWritable
+ * into a {@link StructuredRecord} and emits the StructuredRecord.
+ * </p>
+ * An exception will be thrown if the type of any of the fields does not match the type stated by the user.
+ */
+@Plugin(type = "source")
+@Name("Elasticsearch")
+@Description("CDAP Elasticsearch Batch Sink takes the structured record from the input source " +
+  "and converts it to a JSON, then indexes it in elasticsearch using the index, type, and id specified by the user. " +
+  "The elasticsearch server should be running prior to creating the adapter.")
+public class ElasticsearchSource extends BatchSource<Text, MapWritable, StructuredRecord> {
+  private static final String INDEX_DESC = "The name of the index to query";
+  private static final String TYPE_DESC = "The name of the type where the data is stored.";
+  private static final String QUERY_DESC = "The query to use to import data from the specified index. " +
+    "See elasticsearch for query examples";
+  private static final String HOST_DESC = "The hostname and port for the elasticsearch instance, e.g. localhost:9200";
+  private static final String SCHEMA_DESC = "The schema or mapping of the data in elasticsearch";
+
+  private final ESConfig config;
+  private Schema schema;
+
+  public ElasticsearchSource(ESConfig config) {
+    this.config = config;
+  }
+
+  private String getResource() {
+    return String.format("%s/%s", config.index, config.type);
+  }
+
+  @Override
+  public void initialize(BatchSourceContext context) {
+    schema = parseSchema();
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = context.getHadoopJob();
+    job.setSpeculativeExecution(false);
+    Configuration conf = job.getConfiguration();
+    conf.set("es.nodes", config.hostname);
+    conf.set("es.resource", getResource());
+    conf.set("es.query", config.query);
+    job.setInputFormatClass(EsInputFormat.class);
+    job.setMapOutputKeyClass(Text.class);
+    job.setMapOutputValueClass(MapWritable.class);
+  }
+
+  @Override
+  public void transform(KeyValue<Text, MapWritable> input, Emitter<StructuredRecord> emitter) throws Exception {
+    emitter.emit(convertToRecord(input.getValue(), schema == null ? parseSchema() : schema));
```

Comment:
Won't the schema be initialized in the initialize method? Is there a need for a null check?